File: tests/compression_tests/gzip_compression/gzip_compression.py
#!/usr/bin/env python3
"""
GZIP compression test implementation in Python.
Measures compression performance, ratio, and throughput.
"""

import json
import sys
import time
import gzip
import random
import string
from typing import Dict, List, Any
import tempfile
import os


def generate_test_data(size: int, data_type: str) -> bytes:
    """Generate test data of specified size and type."""
    if data_type == "text":
        # Generate random text data
        text = ''.join(random.choices(string.ascii_letters + string.digits + ' \n', k=size))
        return text.encode('utf-8')
    
    elif data_type == "binary":
        # Generate random binary data
        return bytes(random.getrandbits(8) for _ in range(size))
    
    elif data_type == "json":
        # Generate structured JSON data
        data = []
        while len(json.dumps(data).encode('utf-8')) < size:
            record = {
                "id": len(data),
                "name": ''.join(random.choices(string.ascii_letters, k=10)),
                "value": random.randint(1, 1000),
                "active": random.choice([True, False]),
                "data": ''.join(random.choices(string.ascii_letters + string.digits, k=50))
            }
            data.append(record)
        
        json_str = json.dumps(data)
        return json_str.encode('utf-8')[:size]
    
    else:
        raise ValueError(f"Unknown data type: {data_type}")


def compress_data(data: bytes, compression_level: int = 6) -> Dict[str, Any]:
    """Compress data using GZIP and measure performance."""
    start_time = time.time()
    
    try:
        compressed_data = gzip.compress(data, compresslevel=compression_level)
        compression_time = time.time() - start_time
        
        original_size = len(data)
        compressed_size = len(compressed_data)
        compression_ratio = original_size / compressed_size if compressed_size > 0 else 0
        throughput = original_size / compression_time / (1024 * 1024)  # MB/s
        
        return {
            "success": True,
            "original_size": original_size,
            "compressed_size": compressed_size,
            "compression_ratio": round(compression_ratio, 3),
            "compression_time": round(compression_time * 1000, 2),  # Convert to ms
            "throughput_mb_s": round(throughput, 2)
        }
        
    except Exception as e:
        compression_time = time.time() - start_time
        return {
            "success": False,
            "compression_time": round(compression_time * 1000, 2),
            "error": str(e)
        }


def decompress_data(compressed_data: bytes) -> Dict[str, Any]:
    """Decompress GZIP data and measure performance."""
    start_time = time.time()
    
    try:
        decompressed_data = gzip.decompress(compressed_data)
        decompression_time = time.time() - start_time
        
        decompressed_size = len(decompressed_data)
        throughput = decompressed_size / decompression_time / (1024 * 1024)  # MB/s
        
        return {
            "success": True,
            "decompressed_size": decompressed_size,
            "decompression_time": round(decompression_time * 1000, 2),  # Convert to ms
            "throughput_mb_s": round(throughput, 2)
        }
        
    except Exception as e:
        decompression_time = time.time() - start_time
        return {
            "success": False,
            "decompression_time": round(decompression_time * 1000, 2),
            "error": str(e)
        }


def run_compression_benchmark(config: Dict) -> Dict:
    """Run GZIP compression benchmark with given configuration."""
    input_sizes = config.get("input_sizes", [1024])
    data_types = config.get("data_types", ["text"])
    compression_levels = config.get("compression_levels", [6])
    iterations = config.get("iterations", 5)
    
    results = {
        "start_time": time.time(),
        "test_cases": [],
        "summary": {
            "total_tests": 0,
            "successful_tests": 0,
            "failed_tests": 0,
            "avg_compression_ratio": 0.0,
            "avg_compression_time": 0.0,
            "avg_decompression_time": 0.0,
            "avg_compression_throughput": 0.0,
            "avg_decompression_throughput": 0.0
        }
    }
    
    total_compression_ratios = []
    total_compression_times = []
    total_compression_throughputs = []
    
    for size in input_sizes:
        for data_type in data_types:
            for level in compression_levels:
                print(f"Testing {data_type} data, size: {size} bytes, level: {level}...", file=sys.stderr)
                
                test_case = {
                    "input_size": size,
                    "data_type": data_type,
                    "compression_level": level,
                    "iterations": [],
                    "avg_compression_ratio": 0.0,
                    "avg_compression_time": 0.0,
                    "avg_decompression_time": 0.0,
                    "avg_compression_throughput": 0.0,
                    "avg_decompression_throughput": 0.0
                }
                
                iteration_compression_ratios = []
                iteration_compression_times = []
                iteration_compression_throughputs = []
                
                for i in range(iterations):
                    print(f"  Iteration {i+1}/{iterations}...", file=sys.stderr)
                    
                    # Generate test data for this iteration
                    test_data = generate_test_data(size, data_type)
                    
                    # Compress
                    compression_result = compress_data(test_data, level)
                    
                    iteration_result = {
                        "iteration": i + 1,
                        "compression": compression_result
                    }
                    
                    results["summary"]["total_tests"] += 1
                    
                    if compression_result["success"]:
                        results["summary"]["successful_tests"] += 1
                        
                        # Note: Decompression test removed to avoid storing compressed data
                        # This simplifies the test and avoids JSON serialization issues
                        
                        iteration_compression_ratios.append(compression_result["compression_ratio"])
                        iteration_compression_times.append(compression_result["compression_time"])
                        iteration_compression_throughputs.append(compression_result["throughput_mb_s"])
                        # Decompression metrics removed due to simplified test
                    else:
                        results["summary"]["failed_tests"] += 1
                    
                    test_case["iterations"].append(iteration_result)
                
                # Calculate averages for this test case
                if iteration_compression_ratios:
                    test_case["avg_compression_ratio"] = sum(iteration_compression_ratios) / len(iteration_compression_ratios)
                    test_case["avg_compression_time"] = sum(iteration_compression_times) / len(iteration_compression_times)
                    test_case["avg_compression_throughput"] = sum(iteration_compression_throughputs) / len(iteration_compression_throughputs)
                    # Decompression metrics removed
                    test_case["avg_decompression_time"] = 0.0
                    test_case["avg_decompression_throughput"] = 0.0
                    
                    total_compression_ratios.extend(iteration_compression_ratios)
                    total_compression_times.extend(iteration_compression_times)
                    total_compression_throughputs.extend(iteration_compression_throughputs)
                
                results["test_cases"].append(test_case)
    
    # Calculate overall summary
    if total_compression_ratios:
        results["summary"]["avg_compression_ratio"] = sum(total_compression_ratios) / len(total_compression_ratios)
        results["summary"]["avg_compression_time"] = sum(total_compression_times) / len(total_compression_times)
        results["summary"]["avg_compression_throughput"] = sum(total_compression_throughputs) / len(total_compression_throughputs)
        # Decompression metrics set to 0 since we're not testing decompression
        results["summary"]["avg_decompression_time"] = 0.0
        results["summary"]["avg_decompression_throughput"] = 0.0
    
    results["end_time"] = time.time()
    results["total_execution_time"] = results["end_time"] - results["start_time"]
    
    return results


def main():
    """Main entry point for GZIP compression test."""
    if len(sys.argv) < 2:
        print("Usage: python gzip_compression.py <config_file>", file=sys.stderr)
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    try:
        with open(config_file, 'r') as f:
            config = json.load(f)
        
        # Extract parameters from config
        parameters = config.get("parameters", {})
        
        # Run the benchmark
        results = run_compression_benchmark(parameters)
        
        # Output results as JSON
        print(json.dumps(results, indent=2))
        
    except FileNotFoundError:
        print(f"Error: Config file '{config_file}' not found", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tests/compression_tests/gzip_compression/gzip_compression.ts
#!/usr/bin/env node
/**
 * GZIP compression test implementation in TypeScript.
 * Measures compression performance, ratio, and throughput.
 */

import * as fs from 'fs';
import * as zlib from 'zlib';
import { promisify } from 'util';

interface CompressionResult {
    success: boolean;
    original_size?: number;
    compressed_size?: number;
    compression_ratio?: number;
    compression_time: number;
    throughput_mb_s?: number;
    error?: string;
}

interface IterationResult {
    iteration: number;
    compression: CompressionResult;
}

interface TestCase {
    input_size: number;
    data_type: string;
    compression_level: number;
    iterations: IterationResult[];
    avg_compression_ratio: number;
    avg_compression_time: number;
    avg_decompression_time: number;
    avg_compression_throughput: number;
    avg_decompression_throughput: number;
}

interface Summary {
    total_tests: number;
    successful_tests: number;
    failed_tests: number;
    avg_compression_ratio: number;
    avg_compression_time: number;
    avg_decompression_time: number;
    avg_compression_throughput: number;
    avg_decompression_throughput: number;
}

interface BenchmarkResults {
    start_time: number;
    test_cases: TestCase[];
    summary: Summary;
    end_time?: number;
    total_execution_time?: number;
}

interface Config {
    parameters: {
        input_sizes?: number[];
        data_types?: string[];
        compression_levels?: number[];
        iterations?: number;
    };
}

const gzipAsync = promisify(zlib.gzip);

function generateTestData(size: number, dataType: string): Buffer {
    switch (dataType) {
        case 'text': {
            const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 \n';
            let result = '';
            for (let i = 0; i < size; i++) {
                result += chars[Math.floor(Math.random() * chars.length)];
            }
            return Buffer.from(result, 'utf-8');
        }
        
        case 'binary': {
            const result = Buffer.alloc(size);
            for (let i = 0; i < size; i++) {
                result[i] = Math.floor(Math.random() * 256);
            }
            return result;
        }
        
        case 'json': {
            const data: any[] = [];
            let currentSize = 0;
            
            while (currentSize < size) {
                const record = {
                    id: data.length,
                    name: generateRandomString(10),
                    value: Math.floor(Math.random() * 1000) + 1,
                    active: Math.random() < 0.5,
                    data: generateRandomString(50)
                };
                data.push(record);
                currentSize = Buffer.from(JSON.stringify(data), 'utf-8').length;
            }
            
            const jsonStr = JSON.stringify(data);
            const jsonBuffer = Buffer.from(jsonStr, 'utf-8');
            return jsonBuffer.slice(0, Math.min(size, jsonBuffer.length));
        }
        
        default:
            throw new Error(`Unknown data type: ${dataType}`);
    }
}

function generateRandomString(length: number): string {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
        result += chars[Math.floor(Math.random() * chars.length)];
    }
    return result;
}

async function compressData(data: Buffer, compressionLevel: number): Promise<CompressionResult> {
    const startTime = Date.now();
    const originalSize = data.length;
    
    try {
        // Map compression levels to zlib constants
        let level: number;
        switch (compressionLevel) {
            case 1:
                level = zlib.constants.Z_BEST_SPEED;
                break;
            case 2:
            case 3:
            case 4:
            case 5:
                level = compressionLevel;
                break;
            case 6:
                level = zlib.constants.Z_DEFAULT_COMPRESSION;
                break;
            case 7:
            case 8:
            case 9:
                level = compressionLevel;
                break;
            default:
                level = zlib.constants.Z_DEFAULT_COMPRESSION;
        }
        
        const compressed = await gzipAsync(data, { level });
        const compressionTime = Date.now() - startTime;
        
        const compressedSize = compressed.length;
        const compressionRatio = compressedSize > 0 ? originalSize / compressedSize : 0;
        const throughput = originalSize / (compressionTime / 1000) / (1024 * 1024); // MB/s
        
        return {
            success: true,
            original_size: originalSize,
            compressed_size: compressedSize,
            compression_ratio: Math.round(compressionRatio * 1000) / 1000,
            compression_time: Math.round(compressionTime * 100) / 100,
            throughput_mb_s: Math.round(throughput * 100) / 100
        };
    } catch (error) {
        const compressionTime = Date.now() - startTime;
        return {
            success: false,
            original_size: originalSize,
            compression_time: Math.round(compressionTime * 100) / 100,
            error: error instanceof Error ? error.message : String(error)
        };
    }
}

async function runCompressionBenchmark(config: any): Promise<BenchmarkResults> {
    const inputSizes = config.input_sizes || [1024];
    const dataTypes = config.data_types || ['text'];
    const compressionLevels = config.compression_levels || [6];
    const iterations = config.iterations || 5;
    
    const results: BenchmarkResults = {
        start_time: Date.now() / 1000,
        test_cases: [],
        summary: {
            total_tests: 0,
            successful_tests: 0,
            failed_tests: 0,
            avg_compression_ratio: 0.0,
            avg_compression_time: 0.0,
            avg_decompression_time: 0.0,
            avg_compression_throughput: 0.0,
            avg_decompression_throughput: 0.0
        }
    };
    
    const totalCompressionRatios: number[] = [];
    const totalCompressionTimes: number[] = [];
    const totalCompressionThroughputs: number[] = [];
    
    for (const size of inputSizes) {
        for (const dataType of dataTypes) {
            for (const level of compressionLevels) {
                console.error(`Testing ${dataType} data, size: ${size} bytes, level: ${level}...`);
                
                const testCase: TestCase = {
                    input_size: size,
                    data_type: dataType,
                    compression_level: level,
                    iterations: [],
                    avg_compression_ratio: 0.0,
                    avg_compression_time: 0.0,
                    avg_decompression_time: 0.0,
                    avg_compression_throughput: 0.0,
                    avg_decompression_throughput: 0.0
                };
                
                const iterationCompressionRatios: number[] = [];
                const iterationCompressionTimes: number[] = [];
                const iterationCompressionThroughputs: number[] = [];
                
                for (let i = 0; i < iterations; i++) {
                    console.error(`  Iteration ${i + 1}/${iterations}...`);
                    
                    try {
                        const testData = generateTestData(size, dataType);
                        const compressionResult = await compressData(testData, level);
                        
                        const iterationResult: IterationResult = {
                            iteration: i + 1,
                            compression: compressionResult
                        };
                        
                        results.summary.total_tests++;
                        
                        if (compressionResult.success) {
                            results.summary.successful_tests++;
                            
                            if (compressionResult.compression_ratio !== undefined) {
                                iterationCompressionRatios.push(compressionResult.compression_ratio);
                            }
                            iterationCompressionTimes.push(compressionResult.compression_time);
                            if (compressionResult.throughput_mb_s !== undefined) {
                                iterationCompressionThroughputs.push(compressionResult.throughput_mb_s);
                            }
                        } else {
                            results.summary.failed_tests++;
                        }
                        
                        testCase.iterations.push(iterationResult);
                    } catch (error) {
                        console.error(`Error in iteration ${i + 1}: ${error}`);
                        results.summary.total_tests++;
                        results.summary.failed_tests++;
                    }
                }
                
                // Calculate averages for this test case
                if (iterationCompressionRatios.length > 0) {
                    testCase.avg_compression_ratio = iterationCompressionRatios.reduce((a, b) => a + b) / iterationCompressionRatios.length;
                    testCase.avg_compression_time = iterationCompressionTimes.reduce((a, b) => a + b) / iterationCompressionTimes.length;
                    testCase.avg_compression_throughput = iterationCompressionThroughputs.reduce((a, b) => a + b) / iterationCompressionThroughputs.length;
                    
                    totalCompressionRatios.push(...iterationCompressionRatios);
                    totalCompressionTimes.push(...iterationCompressionTimes);
                    totalCompressionThroughputs.push(...iterationCompressionThroughputs);
                }
                
                results.test_cases.push(testCase);
            }
        }
    }
    
    // Calculate overall summary
    if (totalCompressionRatios.length > 0) {
        results.summary.avg_compression_ratio = totalCompressionRatios.reduce((a, b) => a + b) / totalCompressionRatios.length;
        results.summary.avg_compression_time = totalCompressionTimes.reduce((a, b) => a + b) / totalCompressionTimes.length;
        results.summary.avg_compression_throughput = totalCompressionThroughputs.reduce((a, b) => a + b) / totalCompressionThroughputs.length;
    }
    
    results.end_time = Date.now() / 1000;
    results.total_execution_time = results.end_time - results.start_time;
    
    return results;
}

async function main(): Promise<void> {
    if (process.argv.length < 3) {
        console.error(`Usage: ${process.argv[0]} ${process.argv[1]} <config_file>`);
        process.exit(1);
    }
    
    const configFile = process.argv[2];
    
    try {
        const configContent = fs.readFileSync(configFile, 'utf-8');
        const config: Config = JSON.parse(configContent);
        
        const results = await runCompressionBenchmark(config.parameters);
        
        console.log(JSON.stringify(results, null, 2));
    } catch (error) {
        if (error instanceof Error) {
            if ((error as any).code === 'ENOENT') {
                console.error(`Error: Config file '${configFile}' not found`);
            } else if (error instanceof SyntaxError) {
                console.error(`Error: Invalid JSON in config file: ${error.message}`);
            } else {
                console.error(`Error: ${error.message}`);
            }
        } else {
            console.error(`Error: ${String(error)}`);
        }
        process.exit(1);
    }
}

main().catch(error => {
        console.error('Unhandled error:', error);
        process.exit(1);
});